Playwright: migrate Gutenboarding (happy path) spec.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to migrate the Gutenboarding spec in Selenium, albeit with a few changes.

Key changes:
- Playwright spec focus on the happy path only for now, meaning no meandering back and forth or selecting options that will not be pursued.
- creation of `GutenboardingFlow` object that holds all actions pertaining to the flow.
- addition of helper methods where necessary to implement this flow.

Background details:

Selenium version of the Gutenboarding flow spec meanders occasionally to select options and/or features that will not be applied for the actually created site (which is Free). Such examples include custom domains, changing of language to `es-ES` and others.

The Playwright version of the spec strips out the meandering to focus on the happy path only:
- unique site name is given;
- free `wordpress.com` domain is chosen;
- free design is chosen;
- font pairing is chosen;
- `Plugin` feature is chosen;
- `Free` plan is chosen at the end.

Minor helper methods and tweaks have also been applied to files such as `SidebarComponent` and `DataHelper`.

#### Testing instructions

1. pull the branch
2. `yarn install`
3. `cd test/e2e`
4. `yarn jest specs/specs-playwright/wp-gutenboarding__create.ts`

Ensure that if you run the test spec locally that all steps (including the site deletion flow) completes, otherwise it is best to delete the site manually.

- [x] Playwright mobile passes.
- [x] Playwright desktop passes.


Related to #

worldomonation committed:
Migrate Gutenboarding tests to Playwright.

worldomonation committed:
Simplify `gutenboardingFlow.selectPlan`.

worldomonation committed:
Refactor how the `selectPlan` works in GutenboardingFlow.
Add wait for initializer for Sidebar.

worldomonation committed:
Fix the plan selection process

worldomonation commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -1,6 +1,7 @@
 import phrase from 'asana-phrase';
 import config from 'config';
 import { getViewportName } from './browser-helper';
+import { createTimestamp } from './media-helper';
```
In a future PR I think I will review the methods in DataHelper and MediaHelper.

worldomonation commented on packages/calypso-e2e/src/lib/components/site-select-component.ts:
```diff
@@ -68,4 +56,16 @@ export class SiteSelectComponent {
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async addSite(): Promise< void > {
+		await Promise.all( [
+			this.page.waitForNavigation(),
+			this.page.click( 'a:text("Add New Site")' ),
+		] );
```
I debated where to place this method, but settled on the `SiteSelectComponen`: the class name of the element mentions `site-select`.

dpasque replied:
Hmm, this one definitely is tricky. I don't want to start too much back-and-forth, but I will say that this does feel a little weird to me to be in the SiteSelectComponent. It feels like it should either be its own new component POM, or part of the sidebar.

To me, we seem to define POMs not based on overall functional intention, but actual physical structure in the UI. From that perspective, this is pretty separate from the whole-page site selector component.

worldomonation replied:
I don't think this deserves its own component/page, since it isn't physically or functionally separate from either the Sidebar or SiteSelect. Bit of a stretch, but an argument could be that 'creating a new site will switch the user to that site!'. 

With that said, I am with you that it feels...a bit off. If anything I'd like to see this placed on the SidebarComponent. What do you think?

worldomonation replied:
I've moved it to the SidebarComponent in 6ee92e2dbe94c57cdec6f968e1423e98211f8841.

worldomonation commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -0,0 +1,205 @@
+	 * @param {string} text User-visible text on the button.
+	 * @returns {Promise<void>} No return value.
+	 */
+	async clickButton( text: string ) {
+		await this.page.click( selectors.button( text ) );
```
A generic `clickButton` method to handle clicking of any buttons that can be located uniquely on the page.

dpasque replied:
What do you think of moving this function out of POM classes and to a more generic library function? This has a lot of utility everywhere in POMs and even test scripts, and this could also keep things DRYer (you have to re-define the "button" selector in multiple POMs - those wouldn't be necessary any more!)

worldomonation replied:
Kinda going back and forth on this one personally.

This was never codified in the documentation but IIRC @WunderBart and I discussed this and it was agreed to avoid what had befallen the Selenium e2e suite (lots of helper functions, some redundant, some not used, some not even working properly). Do you recall anything of the sort, @WunderBart?

My opinion is that it is preferable to implement these small helper functions individually in the POMs, at risk of repetitiveness. Reason being:
1. normally we work with pages, but sometimes with frames;
2. 'local' implementation will be better suited to the subtle differences present on that page;

Case in point, in another migration PR I implemented another `clickButton` method but with intricacies specific to that page, which wouldn't have worked with how `Gutenboarding` pages work.

dpasque replied:
> avoid what had befallen the Selenium e2e suite (lots of helper functions, some redundant, some not used, some not even working properly)

This makes a lot of sense to me!

> My opinion is that it is preferable to implement these small helper functions individually in the POMs, at risk of repetitiveness.

Totally down for this. I think it fits how we're trying to grow well!

In that case, should we make this method private?

worldomonation replied:
It should not be private as it's used by the test spec to perform actions on the page. I don't think it was valuable creating a method for each individual button action (eg. clickContinue, clickCancel, clickXYZ) as it only clutters the class. Do you think that's preferable for clarity?

dpasque replied:
Ah whoops, I missed that, sorry! Hmm, no, I've actually really come around on this. I think this is good as is!

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -54,7 +54,7 @@ export class GutenbergEditorPage {
 	async waitUntilLoaded(): Promise< Frame > {
 		const frame = await this.getEditorFrame();
-		await this.page.waitForLoadState( 'networkidle' );
+		await this.page.waitForLoadState( 'networkidle', { timeout: 60000 } );
```
I noticed the iframe'd editor may take a long time to appear. The extended timeout is to help guard against an otherwise normally loading page.

worldomonation committed:
Use updated name for method

dpasque commented on packages/calypso-e2e/src/lib/pages/general-settings-page.ts:
```diff
@@ -0,0 +1,52 @@
+			this.page.click( selectors.button( 'Delete this site' ) ),
+		] );
+
+		// Redirected page should no longer reference the deleted site in the URL.
```
I'm not entire sure the answer to my own question, but it got me thinking: should these kind of assertions that the deletion was successful be left to the tests that consume this action? It feels a little weird to have an assertion in a POM class method, but it's probably also fine.

If we leave this here, I think we definitely should add a custom message to the `assert.ok()` call - something like `"Navigation away from deleted site did not occur. The site deletion may have failed."`

worldomonation replied:
I recall there was a similar thread of discussion some weeks ago, when you first joined the team. 

I do not recall what the conclusion of that discussion was. However, assertions in page objects get enough searches that I get a good number of results when I query this.

For the most part, results I see are in line with what you say, that assertions or checks should be done in the tests themselves, not the page objects. [1][2]

With that said, `validateXYZ` methods are also technically assertions, just framed as 'checking that page is showing xyz', so there's that too. 

Ultimately I think sometimes an exception can be made and maybe this one is it, since deleting a site is a pretty big deal - I felt that having the checks built-in to the page object would ensure better results vs. leaving it to the test writer to deal with.

[1] https://stackoverflow.com/questions/11126806/page-object-model-why-not-include-assertions-in-page-methods
[2] https://medium.com/ranorex-webtestit/5-tips-to-keep-your-page-objects-maintainable-3686bd539759#91a1

worldomonation replied:
You know what, after some more thinking I've come around on it. I'll remove the assertion in the object and expect the caller to verify this.

worldomonation replied:
Fixed in 32357f3ed0c917646b87a95c963f9109d720b478.

dpasque replied:
Looks good! But I think you also make pretty compelling points - sometimes if that logic is coupled to the details of the page, it might make sense to have assertions in the POMs themselves (like you mention in the `validateXYZ`). I'll keep that in mind as I do future migrations!

dpasque commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -0,0 +1,205 @@
+		`.domain-picker__suggestion-item-name span:has-text("${ name }${ tld }")`,
+
+	// Design
+	design: ( name: string ) => `button[data-e2e-button="freeOption"] span:text("${ name }")`,
```
Nit: we really are most interested in the button here, right? Would it make sense to use the "has-text" pseudo-selector here?

`button[data-e2e-button="freeOption"]:has-text("${ name }")`

worldomonation replied:
Good call, thanks!

worldomonation replied:
Fixed in Fixed in 32357f3ed0c917646b87a95c963f9109d720b478.

dpasque commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -0,0 +1,205 @@
+		await this.page.click( selectors.fontPairingButton( viewportName, name ) );
+
+		if ( viewportName === 'desktop' ) {
+			await this.page.waitForSelector( `button.is-selected span:text("${ name }")` );
+		} else {
+			await this.page.waitForSelector(
+				`button.is-selected-dropdown-option span:text("${ name }")`,
+				{ state: 'hidden' }
+			);
+		}
+	}
+
+	/**
+	 * Given an array of feature strings, selects matching features.
+	 *
+	 * @param {Features} features Array of features to select.
+	 */
+	async selectFeatures( features: Features[] ): Promise< void > {
+		for await ( const feature of features ) {
+			// Click on the feature to select, then make sure it is in fact selected by checking
+			// whether the attribute `is-selected` is present for the button with that feature.
+			await this.page.click( `${ selectors.featureItem } :text-is("${ feature }")` );
+			await this.page.waitForSelector(
+				`${ selectors.featureItem }.is-selected :text-is("${ feature }")`
```
I personally think this is okay, but would it be better for consistency to include all these selectors in the selectors constant at the top of the module?

worldomonation replied:
I tend to shy away from creating variations of selectors, so I think I'd rather refactor this to obtain the target element then check whether the `classList` contains a `is-selected` instead.

dpasque replied:
> I tend to shy away from creating variations of selectors

That makes sense to me. I think we should add that to the Playwright docs at some point then to codify that! I.e., explain that if you're iterating on a base selector, it is okay to do that in-line.

I don't think you need to change this then.

dpasque commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -0,0 +1,205 @@
+		// Iterate through the top-level container for each plan.
+		// Then, extract name of the plan and if it finds a match with the desired plan,
+		// click on the `Select` button within the container.
+		for await ( const plan of plans ) {
+			const planName = await plan
+				.waitForSelector( selectors.planName )
+				.then( ( el ) => el.innerText() );
+
+			console.log( planName );
+			if ( planName === name ) {
+				const button = await plan.waitForSelector( selectors.button( 'Select' ) );
+				await Promise.all( [ this.page.waitForNavigation(), button.click() ] );
+				return;
+			}
+		}
+	}
```
I think we can do a much simpler implementation of this with just a slightly more fancy selector!

Here's an example of what I tried and worked, but feel free to iterate/clean-up :) 

```typescript
async selectPlan( name: Plans ): Promise< void > {
		// First, expand the accordion.
		await this.page.click( ':text-is("Show all plans")' );
		await Promise.all( [
			this.page.waitForNavigation(),
			this.page.click(
				`.plans-accordion-item:has(.plans-accordion-item__heading:has-text("${ name }")) button:has-text("Select")`
			),
		] );

		// const plans = await this.page.$$( selectors.planItem );

		// // Iterate through the top-level container for each plan.
		// // Then, extract name of the plan and if it finds a match with the desired plan,
		// // click on the `Select` button within the container.
		// for await ( const plan of plans ) {
		// 	const planName = await plan
		// 		.waitForSelector( selectors.planName )
		// 		.then( ( el ) => el.innerText() );

		// 	console.log( planName );
		// 	if ( planName === name ) {
		// 		const button = await plan.waitForSelector( selectors.button( 'Select' ) );
		// 		await Promise.all( [ this.page.waitForNavigation(), button.click() ] );
		// 		return;
		// 	}
		// }
	}
```

worldomonation replied:
Thanks for the suggestion, I'll give your selector a try.

worldomonation replied:
Interestingly, the suggested selector matches more than 1 item:

![image](https://user-images.githubusercontent.com/6549265/130537237-5f416313-87e5-480e-afa7-a2e8ce7d910c.png)

It is matching on the text that states `Pick a free domain (1 year)`.

It looks like that with the accordion expanded, it matches the 'Free' plan first, but I wonder if this would be the case if the order was shifted in the future.

A note, I struggled with the selector for the business plan for the longest time while working on this PR. I'm okay with accepting a selector that matches multiple items, but personally I think a traditional for/if combination is easier to follow.

worldomonation replied:
I've changed a portion of the selector to this and it seems to filter out the false match:
```
.plans-accordion-item:has(.plans-accordion-item__name:has-text("${name}")) button:has-text("Select")'
```

worldomonation replied:
Fixed in 32357f3ed0c917646b87a95c963f9109d720b478.

dpasque replied:
> personally I think a traditional for/if combination is easier to follow

Interesting! That's helpful to hear, because I actually had the opposite gut reaction when I first read this! This might be worth discussing as a group later as there might be a style-guide point to make here. Complicated selection is inevitable, is it better to blow that out into DOM iterating, or condense into a single selector?

worldomonation replied:
I agree that it's worth having a discussion over. I prefer a for loop as Playwright makes it easy to select all matching top-level items. By iterating on the top-level items it's easy to limit search scope to the child elements.

dpasque commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -0,0 +1,205 @@
+	 * @param {string} param0.tld TLD portion of the domain.
+	 * @returns {Promise<string>} Selected domain.
+	 */
+	async selectDomain( { name, tld }: { name: string; tld: string } ): Promise< string > {
```
Definite nit here: but I'm not entirely sure why we need to capture the name and tld as separate parameters here - it looks like we just concatenate them together later. In that case, would it just be better to have the user pass in exactly what domain they want to find? That way there's no risk they accidentally pass in "com" instead of ".com". In that case, we could also take out the code that returns the `innerText` of the found element - we either successfully find the exact URL they are looking for, or we don't!

Alternatively, I think capturing both separately makes a lot of sense if we want to do partial matching (like just a key word that needs to be available in the domain name). But in that case, we should probably use the [Playwright Regex based text selectors](https://playwright.dev/docs/selectors#text-selector) to find something that contains the keyword for the domain:

```typescript
	domainSuggestionSpan: ( name: string, tld: string ) =>
		`.domain-picker__suggestion-item-name :text-matches(".*${ name }.*\\.${ tld }", "i")`,
```

worldomonation replied:
This is a fair point and I had the following use case in mind when I did this:

1. user searches for `mycoolblog`
2. user selects `mycoolblog.com`.
3. user changes mind and returns to the domain search step.
4. user searches for `mysecondcoolblog.com`.

In this scenario, two domains with a `.com` TLDs are shown on page. 
I *think* I was trying to implement something flexible with this, but the end result didn't end up being like that. 
So I think it's preferable to default to a simpler implementation, to request user to enter the exact domain they would like.

worldomonation replied:
Fixed in 32357f3ed0c917646b87a95c963f9109d720b478.

dpasque commented on test/e2e/specs/specs-playwright/wp-gutenboarding__create.ts:
```diff
@@ -0,0 +1,86 @@
+		await gutenbergEditorPage.waitUntilLoaded();
+	} );
+
+	it( `Delete created site`, async function () {
```
I really like having this just as part of the overall test flow! It accomplishes testing something meaningful and also doing data clean-up at the same time.

worldomonation replied:
Interestingly, this was originally implemented as `afterAll` hook, but that has a drawback that it isn't captured on screen as part of the normal test run. So if something goes wrong in the delete site portion on CI, it is difficult to tell what happened.

dpasque commented on test/e2e/specs/specs-playwright/wp-gutenboarding__create.ts:
```diff
@@ -0,0 +1,86 @@
+		try {
+			await generalSettingsPage.deleteSite( siteURL );
+		} catch ( e ) {
+			console.error( `Deleting ${ siteURL } failed, manual intervention needed.` );
```
Out of curiosity: is manual intervention really needed here? I don't think this will disrupt future runs - it will just orphan a test site. That's already something we have a ton of, and I think it makes more sense to try to clean those up with some background job that putting the onus on a test runner (who might be hesitant to then re-run the test)

worldomonation replied:
It will orphan the site - that is correct, but the buildup of orphaned site that belongs to the test user is what ended up slowing down the said test user considerably, so I would prefer that anyone investigating and seeing the error message deal with it.

Probably won't happen, though 😆

worldomonation replied:
I've adjusted the error message so it doesn't call for manual intervention now.

worldomonation replied:
Fixed in 32357f3ed0c917646b87a95c963f9109d720b478.

dpasque reviewed (changes requested):
Overall the test design and proposed changes look fantastic! Got a few comments for your review!

worldomonation committed:
PR comments:
- change text shown when site deletion failed.
- remove assertion from site deletion flow.
- adjust selectors for selecting plan.
- adjust method signature for select domain.

worldomonation committed:
Remove extraneous apostrophe from selector

worldomonation committed:
Move addSite to SidebarComponent.

worldomonation committed:
Fix how expect is used.

dpasque commented on packages/calypso-e2e/src/lib/components/site-select-component.ts:
```diff
@@ -15,14 +14,18 @@ export class SiteSelectComponent {
 	 * Creates an instance of the Site Select component.
 	 *
+	 * The site selector can be shown on either one of the following:
+	 * 	- in a pseudo-sidebar, replacing the navigation sidebar.
```
Small nit - but we should pull this line out now!

worldomonation replied:
No, it should stay - the site selector _does_ have two forms on Calypso as mentioned:

'pseudo-sidebar'
![image](https://user-images.githubusercontent.com/6549265/130655181-15b38807-d1da-4d35-9ff3-82877064ed39.png)

standalone component 
![image](https://user-images.githubusercontent.com/6549265/130655233-60f9bca5-dfa5-4974-b925-abda459a6f36.png)

worldomonation replied:
I've checked that this component works for both forms of the `SiteSelectoComponent`, so the comment does apply here.

dpasque commented on packages/calypso-e2e/src/lib/flows/gutenboarding-flow.ts:
```diff
@@ -0,0 +1,201 @@
+		await Promise.all( [
+			this.page.waitForNavigation(),
+			this.page.click(
+				`.plans-accordion-item:has(.plans-accordion-item__name:has-text("${ name }")) button:has-text("Select")`
```
Small nit (and my fault!): but this should probably go in the top constant selectors as it is not an iteration on another selector

worldomonation replied:
Fixed in 80deae5a8cf3c3e1b5f9cb602870245e968d8ab2

dpasque reviewed (approved):
All the changes look great, thank you! 😄  Have a few small style/doc nits, but approving as they are not major at all.

worldomonation committed:
Move selector up to the selectors object.
Temporarily disable the last three steps leading to site creation, to reduce burden on production database.

worldomonation commented:
Just a heads up @dpasque that I've disabled the actual site creation and deletion part of the test, because of the conversation with Systems. I'd like to see this test turned on in full for `trunk` branches only.

dpasque commented:
@worldomonation Last changes LGTM, and the skipping for now too. It looks like there's a conflict, but after that, should be good to 🚢
